File: csharp/Tennis/TennisGame1.cs
using System.Diagnostics;

namespace Tennis
{
    public class TennisGame1 (string player1Name, string player2Name) : ITennisGame
{
    private Player player1 = new(player1Name);
    private Player player2 = new(player2Name);

        public void WonPoint(string playerName)
        {
            if (playerName == player1.Name)
                player1.Score += 1;
            else if (playerName == player1.Name)
                player2.Score += 1;
            else throw new UnreachableException("Bad player name");
        }

        public string GetScore()
        {
            if (player1.Score == player2.Score) return GetTieScoreTerm();
            
            if (player1.Score >= 4 || player2.Score >= 4)
            {
                var scoreDifference = player1.Score - player2.Score;
                return GetLateGameScore(scoreDifference);
            }
            
            string scoreTerm1 = GetRegularScoreTerm(player1.Score);
            string scoreTerm2 = GetRegularScoreTerm(player2.Score);
            return $"{scoreTerm1}-{scoreTerm2}";
        }

        private string GetTieScoreTerm()
        {
            return player1.Score switch
            {
                0 => "Love-All",
                1 => "Fifteen-All",
                2 => "Thirty-All",
                _ => "Deuce",
            };
        }

        private string GetLateGameScore(int scoreDifference)
        {
            return scoreDifference switch
            {
                1 => $"Advantage {player1.Name}",
                -1 => $"Advantage {player2.Name}",
                >= 2 => $"Win for {player1.Name}",
                _ => $"Win for {player2.Name}",
            };
        }

        private static string GetRegularScoreTerm(int score)
        {
            return score switch
            {
                0 => "Love",
                1 => "Fifteen",
                2 => "Thirty",
                _ => "Forty",
            };
        }
    }
}

File: typescript-jest/src/TennisGame1.ts
import { TennisGame } from './TennisGame';

interface Player {
  name: string;
  score: number;
}

export class TennisGame1 implements TennisGame {
  private playerList: Player[] = [];

  constructor(player1Name: string, player2Name: string) {
    this.playerList.push({name: player1Name, score: 0});
    this.playerList.push({name: player2Name, score: 0});
  }

  wonPoint(playerName: string): void {
    const player = this.playerList.find(_ => _.name === playerName);
    if (player) {
      player.score += 1;
    }
  }

  getScore(): string {
    const deltaResult = this.playerList[0].score - this.playerList[1].score;

    if (this.isScoreEqual(deltaResult)) {
      return this.getEqualScoreStr();
    } else if (this.isWin(deltaResult)) {
      return deltaResult > 0 ? 'Win for player1' : 'Win for player2';
    } else if (this.isMatchPoint(deltaResult)) {
      return deltaResult > 0 ? 'Advantage player1' : 'Advantage player2';
    } else {
      return this.getCurrentScoreText();
    }
  }

  private isWin(deltaResult: number) {
    return (this.playerList[0].score >= 4 || this.playerList[1].score >= 4) && Math.abs(deltaResult) > 1;
  }

  private isMatchPoint(deltaResult: number) {
    return (this.playerList[0].score >= 4 || this.playerList[1].score >= 4) && Math.abs(deltaResult) == 1;
  }

  private isScoreEqual(deltaResult: number) {
    return deltaResult === 0;
  }

  private getEqualScoreStr() {
    switch (this.playerList[0].score) {
      case 0:
        return 'Love-All';
      case 1:
        return 'Fifteen-All';
      case 2:
        return 'Thirty-All';
      default:
        return 'Deuce';
    }
  }

  private getCurrentScoreText() {
    return this.getScoreText(this.playerList[0].score) + '-' + this.getScoreText(this.playerList[1].score);
  }

  private getScoreText(tempScore: number) {
    switch (tempScore) {
      case 0:
        return 'Love';
      case 1:
        return  'Fifteen';
      case 2:
        return  'Thirty';
      case 3:
        return  'Forty';
      default:
        return '';
    }
  }

}
